Mathlib/CategoryTheory/MorphismProperty/Composition.lean

```diff
@@ -198,6 +198,90 @@ instance inf {P Q : MorphismProperty C} [P.IsMultiplicative] [Q.IsMultiplicative
 
 end IsMultiplicative
 
+/-- Given a morphism property `W`, the `MultiplicativeClosure W` is the smallest
+multiplicative property greater or equal than `W`. -/
+inductive multiplicativeClosure (W : MorphismProperty C) : MorphismProperty C
+  | of {x y : C} (f : x ⟶ y) (hf : W f) : multiplicativeClosure W f
+  | id (x : C) : multiplicativeClosure W (𝟙 x)
+  | comp_of {x y z : C} (f : x ⟶ y) (g : y ⟶ z) (hf : multiplicativeClosure W f) (hg : W g) :
+    multiplicativeClosure W (f ≫ g)
+
+/-- A variant of MultiplicativeClosure in which compositions are taken on the left rather than
+on the right. It is not intended to be used directly, and one should rather access this via
+`multiplicativeClosure.eq_prime` in cases where the inductive principle of this variant is
+needed. -/
+inductive multiplicativeClosure' (W : MorphismProperty C) : MorphismProperty C
+  | of {x y : C} (f : x ⟶ y) (hf : W f) : multiplicativeClosure' W f
+  | id (x : C) : multiplicativeClosure' W (𝟙 x)
+  | of_comp {x y z : C} (f : x ⟶ y) (g : y ⟶ z) (hf : W f) (hg : multiplicativeClosure' W g) :
+    multiplicativeClosure' W (f ≫ g)
+
+namespace multiplicativeClosure
+
+variable {W : MorphismProperty C}
+
+/-- `MultiplicativeClosure W` is multiplicative. -/
+instance : IsMultiplicative W.multiplicativeClosure where
+  id_mem x := .id x
+  comp_mem f g hf hg := by
+    induction hg with
+    | of _ hf₀ => exact .comp_of f _ hf hf₀
+    | id _ => rwa [Category.comp_id]
+    | comp_of f' g hf' hg h_rec =>
+      rw [← Category.assoc]
+      exact .comp_of (f ≫ f') g (h_rec f hf) hg
+
+/-- The multiplicative closure is greater or equal than the original property. -/
+lemma self_le : W ≤ W.multiplicativeClosure := fun {_ _} _ hf ↦ .of _ hf
```

Comment:
I would suggest moving lemmas in the `MorphismProperty` namespace, e.g. `MorphismProperty.le_multiplicativeClosure`.